zerver/actions/message_flags.py

```diff
@@ -315,8 +317,18 @@ def do_update_message_flags(
             )
 
         if operation == "add":
+            messages = [
+                message_id
+                for message_id in messages
+                if ~um_flags.get(message_id, DEFAULT_HISTORICAL_FLAGS) & flagattr != 0
+            ]
             count = query.update(flags=F("flags").bitor(flagattr))
         elif operation == "remove":
+            messages = [
+                message_id
+                for message_id in messages
+                if um_flags.get(message_id, DEFAULT_HISTORICAL_FLAGS) & flagattr != 0
+            ]
             count = query.update(flags=F("flags").bitand(~flagattr))
```

Comment:
Just making a note that I checked that we're only changing the contents of the event, not the (silly) behavior of the endpoint itself where we return the set of messages the client sent us.